### ISSUE TITLE
-e ignore and -e abort for more error-handling options

### ISSUE THREAD

**Issue by BoPeng** [vatlab/sos#1266]
https://www.gnu.org/software/make/manual/html_node/Errors.html

We now have -k (#1265 ) for `--keep-going`. GNU Make has an similar option that we can potentially borrow.We could also allow option `allow_errors` in more places .

**Comment by gaow** [comment]
Not sure how that's going to differ from `-k` in essence; other than possibly it is not going to do the atomic write that removes output, if exists & return non-zero status? If so, the behavior is not good ...

**Comment by BoPeng** [comment]
I found a possible application of option. In my data process pipeline, sometimes different parameters are given to skip parts of the pipeline. This results in datasets with "holes" here and there, causing difficulties in the "summarize" step. Although something like

```
[default]
if fileA.exists():
   sos_run('process_fileA')
if fileB.exists():
   sos_run('process_fileB')

[process_fileA]
[process_fileB]
```
will work, the need to figure out what `file_A` should be can be a burden, and it might be easier just to run the pipeline that runs through when `process_fileA` and/or `process_fileB` fails.

**Comment by gaow** [comment]
Okay I'm digging up this old open ticket to continue discussion here for error handling behavior. This is currently an issue with running large benchmark workflows when only a few substeps go wrong. We need two behavior: 

1. an `--ignore-errors` behavior that creates dummy outputs eg with `zapped` extension, for example, for substeps that contains the error. That case the workflow will always run successfully but when we see `zapped` files there we know there was an error (and of course errors will be documented into `stderr` files if they are specified). Here I discuss with a toy example:
2. some behavior to stop a step immediately if any substep fails. I think it was previously the default behavior but we changed it to currently wait for completion of the entire step.

@BoPeng do you think these are reasonable requests that we currently dont have available? Command interface would be a next thing to worry about but I would like to make sure I'm correct about current status.

Here is a toy example we can use to test out the behavior:

```
[global]
parameter: sec = 2
parameter: n = 5
reps = [i for i in range(n)]

[a]
input: for_each = 'reps'
output: f'a{_reps}.txt'
if _index == 3:
	raise ValueError 
else:
	run(f'sleep {sec} && echo a{_index} > {_output}')
[b]
input: for_each = 'reps'
output: f'b{_reps}.txt'
run(f'sleep {sec * 3} && echo b{_index} > {_output}')


[default_1]
depends: sos_step('a')
input: None
output: [f'a{i}.txt' for i in range(n)]

[default_2]
depends: sos_step('b')
input: None
output: [f'b{i}.txt' for i in range(n)]


[clean]
files = [f'a{i}.txt' for i in range(n)] + [f'b{i}.txt' for i in range(n)]
bash: expand = True
	rm -f {' '.join(files)}
```

## Current behaviors

Default behavior:

```
sos run keep-going.sos clean && sos run keep-going.sos -j10
```

output:

```
...
WARNING: Substep a (id=c32cfc1964fed41a, index=3) returns an error. Terminating step after completing 3 submitted substeps.
...
ValueError: 
[default]: 2 pending steps: default_1, default_2
```

Keep going behavior:

```
sos run keep-going.sos clean && sos run keep-going.sos -j 10 -k
```

output:

```
...
WARNING: Substep a (id=f4eec23b175fe854, index=3) returns an error.
...
[default]: 1 pending step: default_1
```

These behavior are as expected and are very nice.

**Comment by BoPeng** [comment]
So is `-k` good enough for your application? To summarize

1. By default failure of a substep will terminate the step, and the workflow **after other running steps are completed**.
2. keep-going means that we observe an error, we keep going until all possible (other) steps are completed. **new steps will be started**, but in the end we report error for the entire `sos run -k`.
3. ignore-error was not implemented, but the idea is that errors are more or less expected and the workflow should continue, and in the end we **report success** for the entire `sos run`, unless the ignored error caused some other steps to fail.

I am not sure how useful `ignore error` is, but in reality, especially for simulations, some percentage of substeps are expected to fail (e.g. with unrealistic simulation parameter). If a global ignore error is too risky, perhaps 

```
[step: ignore_error=True]
```
could be used to indicate the case for particular steps.

**Comment by gaow** [comment]
> So is -k good enough for your application?

`-k` is not good enough, as you can see 4 out of 5 of the output files for `default_1` would have been valid if the workflow can keep going without worrying about failed substeps. As my first proposal in the post above outlines, I would expect to see no error from SoS and the output would be something like `a1.txt`, `a3.txt.zapped`, ... `a5.txt` , and additionally a warning providing some summary of failures. Currently with `-k` there are still errors and the workflow will stop.

> By default failure of a substep will terminate the step

Really? I dont think this is the default behavor now, as you can see in the warning message in my post above, `Terminating step after completing 3 submitted substeps.`. That is why I argue in the 2nd proposal in my post above that it would be nice to have some way to terminate the step immediately as soon as one substep fails.

>  unless the ignored error caused some other steps to fail.

But I was talking about at substep level ignore. Imaging you have a very complicated workflow where the first step has 10K substeps, but only (randomly) 50 of them dont work, for various reasons. What I would hope to see is some way to run the workflow with the 2nd step taking those (10K - 50) not failed runs and continue, rather than having to require me fixing each of those 50 failures before being able to get any meaningful results. Of course in order to do so, I expect that we have to create dummy output files like `.zapped` to satisfy input / output dependencies. 
>

**Comment by BoPeng** [comment]
>  I expect that we have to create dummy output files like .zapped to satisfy input / output dependencies.

`zapped` is for real file that was zapped, not for non-existing files. As I have said, `keep going` only keep going the rest of the workflow, and the step with failed substeps still fails and the entire workflow still fails.

What you are asking is a lot like the `ignore error` behavior where the errors are ignored. How to handle missing `_output` is a problem but I could prefer empty `step_output` over `zapped` placeholders. This will make the following step a bit more difficult but perhaps specifying

```
input: group_by=1
```
again could discard empty `_input` from the `step_input` so that you can process `step_input` normally?

**Comment by BoPeng** [comment]
What I was proposing is something like

```
[step_10: ignore_error]
input: for_each=dict(i=range(10))
output: f'test_{i}'.txt

fail_if(i==3, 'failed')
_output.touch()

[step_20]
input: group=1
# continue processing with reduced number of substeps

[step_30]
input: group='all'
print(f'Number of successful simulations are {len(step_input)}')
```

**Comment by gaow** [comment]
> zapped is for real file that was zapped, not for non-existing files. As I have said, keep going only keep going the rest of the workflow, and the step with failed substeps still fails and the entire workflow still fails.

Sure, I understand the current behavior and I showed and agreed in my first post that current behavior is nice. I also only use `zapped` to refer to a dummy input/output file and a possible mechanism to make this behavior happen. We definitely can and want to use a different file extension for it.

Yes I wanted to discuss `--ignore-errors ` behavior and for my user case as you see, a global `--ignore-errors` behavior would be fine. 

> How to handle missing _output is a problem but I could prefer empty step_output over zapped placeholders.

sure! I was thinking of having some file extensions like zapped so the next step will know about it and ignore it by setting the corresponding output also to an empty file with certain extension.

> again could discard empty _input from the step_input so that you can process step_input normally?

It would not work for my case (the way I implemented DSC benchmark workflow) because all files names are pre-determined and a step will explicitly look for all specified input names. So your proposal above would not work for that case.

**Comment by BoPeng** [comment]
So what you are saying is that in the case of 

```
[step_10]
input: for_each=dict(i=range(10))
output: f'test_{i}'.txt

fail_if(i==3, 'failed')
_output.touch()

[step_20]
# continue processing with reduced number of substeps

[step_30]
input: group='all'
print(f'Number of successful simulations are {len(step_input)}')
```

`step_20` will just skip the substep with empty `_input` because it is an error caused by the upstream step?

**Comment by gaow** [comment]
Yes exactly. But step_20 will explicitly require `input` and `output`, even though it decides it can process with reduced number of substeps based on number of failures occurred upstream. My toy example in the first post of this ticket suggests that behavior -- you see there I do expect to see complete list of files in `output` of `default`, not some dynamically determined subsets from successful runs.

**Comment by BoPeng** [comment]
OK, `-i` is available but many would consider it as `--input`. Not sure what else sounds like ignore but I would open to something like

```
-e keep-going
-e ignore
```
which is more readable than `-k` and `-i`.

**Comment by gaow** [comment]
I agree `-e` is reasonable. And just a reminder that we have not discussed the behavior to kill the current running step immediately when a substep fails. So perhaps we need:

```
-e keep-going
-e ignore
-e kill
```
?

**Comment by gaow** [comment]
`-e kill` will be useful for say if I have a typo in my script that will ruin all 10K runs ... we want it to terminal sooner than later. But alternatively I think we can live without using `-e kill` but instead of 

```
WARNING: Substep a (id=c32cfc1964fed41a, index=3) returns an error. Terminating step after completing 3 submitted substeps.
```

we tell the users where to look for the error? eg, where stderr file is stored. That case users can examine it while other substeps still runs, and decide if to wait until later or to `ctrl-c` it immediately.

**Comment by BoPeng** [comment]
ok, they are on my radar now, will be done after `-r host` with templates.

**Comment by gaow** [comment]
Great -- i want to bring up a point that might matter: sometimes a failed substep is not necessarily identified by a non-zero return code. It could be, say out of walltime or memory, or simply randomly killed on the cluster. So maybe we cannot merely rely on the conventional error handling mechanism (eg, `try ... catch`) to creating empty output files.

**Comment by gaow** [comment]
@BoPeng no rush for the next couple of days, but I'm adding a `soon` label to this ticket if it can be prioritized. Currently there are a few of my simulation runs that simply cannot complete due to resource limitations. But failure to complete them on the cluster currently disrupts workflow that follow those failed steps. I think `-e ignore` is the only solution to the problem.

**Comment by BoPeng** [comment]
Yes. It is hard to work on this in Denver with ski on. :-)

**Comment by BoPeng** [comment]
`-e abort` is missing. `-e ignore` is being implemented for various cases.

```
import time

[10]
input: for_each=dict(i=range(10)), concurrent=False
output: f'test_{i}.txt'

_output.touch()

fail_if(_index == 5, 'fail at 5')

[20]
output: _input.with_suffix('.bak')
_output.touch()
```
is the test case for failure in one of the substeps.

**Comment by gaow** [comment]
@BoPeng here is a test for `-e ignore` with dynamic input -- that is, previously failed step will not generate some inputs the next step is expecting. The desired behavior is to be able to ignore all these problems with a summary of ignored errors at the very end.

[issue_1266.tar.gz](https://github.com/vatlab/sos/files/3913028/issue_1266.tar.gz)

To reproduce, simply:

```
sos run test_run.sos
```

**Comment by BoPeng** [comment]
https://vatlab.github.io/sos-docs/doc/user_guide/error_handling.html tries to document the new features.

It appears that `keep-going` could try to execute the next step, but then it would be the same as `ignore` just with different final return code of the workflow.

**Comment by BoPeng** [comment]
@gaow Frankly I do not know the intended behavior for subworkflows. In your example, 

1. `simulate` is called but only generates 4 out of 5 needed targets
2. `a_simulate` is expecting 5 from `sos_run('simulate')`. It only got 4, so `a_simulate` is reported to be failed.
3. The failure of `a_simulate` makes `sos_step('a_simulate')` unavailable, so `a_test` cannot continue. The same goes to `default`.

**Comment by gaow** [comment]
> until you have a clear logic as what error should be ignored at the step level

Not sure if this "logic" is clear --- I need a step to success and generate all the output files that are explicitly required by the next step using `dynamic`. So 

1. `a_simulate` should not return failed status.
2. `a_test` should run even some of its input are not available because of failed output from previous step.

That was why I thought making some fake input like `zapped` (with a different name) can help.

> I do not know the intended behavior for subworkflows

Well, perhaps subworkflow is not the best way to do it, but the idea (or the goal) is to have prespecfied input and output file names. My workflow steps are supposed to generate them. If some of those files are not generated, I am hoping they can be replaced with dummy files to keep the process going. I'm definitely open to other possibilities to achieve it. I'll DM you for additional details.

**Comment by BoPeng** [comment]
Behavior of nested workflow is improved, but I am not sure if the behavior is optimal.

```
(sos) [bpeng1@BCBMF5KWK0F8F69:~/sos/sos/issue (master *)]$ sos run test_run.sos -e ignore -s force
INFO: Running simulate in pipeline #1: 
INFO: simulate in pipeline #1 is completed (pending nested workflow).
INFO: Running module simulate: 
INFO: simulate (index=0) is completed.
INFO: simulate (index=1) is completed.
INFO: simulate (index=2) is completed.
INFO: simulate (index=3) is completed.
WARNING: Ignoring error from substep simulate (id=b7af26066c5f1731, index=4): 
Failed to execute Rscript .sos/simulate_4_5ddd04ef.R
exitcode=1, workdir=/Users/bpeng1/sos/sos/issue, stderr=test/simulate/simulate_5.stderr
---------------------------------------------------------------------------.
INFO: simulate output:   test/simulate/simulate_1.rds test/simulate/simulate_2.rds... (5 items in 5 groups)
INFO: a_simulate output:   test/simulate/simulate_1.rds test/simulate/simulate_2.rds... (5 items)
WARNING: Output target test/simulate/simulate_5.rds does not exist after the completion of step a_simulate
INFO: Running test in pipeline #1: 
INFO: test in pipeline #1 is completed (pending nested workflow).
INFO: Running module test: 
WARNING: Failed to regenerate or resolve test/simulate/simulate_5.rds.
WARNING: Error from step test is ignored: [test]: Target unavailable: test/simulate/simulate_5.rds
INFO: a_test output:   test/test/simulate_1_test_1.rds test/test/simulate_2_test_1.rds... (5 items)
INFO: Running default: 
INFO: default is completed.
INFO: default output:   test/test/simulate_1_test_1.rds test/test/simulate_2_test_1.rds... (5 items)
INFO: Workflow default (ID=0825ced33264897e) is executed successfully with 4 completed steps and 8 completed substeps.
```

**Comment by gaow** [comment]
Well, the workflow seemingly completes but I expected to see 4 outputs in both `simulate` and `test` folder under `test`. Currently, 

```
$ find . -name "*.rds"
./simulate/simulate_4.rds
./simulate/simulate_2.rds
./simulate/simulate_3.rds
./simulate/simulate_1.rds
```

only has results for `simulate`. `simulate_5.rds` is missing. The desired behavior, if possible, is to let the `test` step still work on the 4 `simulate_*.rds` and generate 4 `test_*.rds`.

**Comment by BoPeng** [comment]
This is because the `test` step cannot run at all due to missing one of the input files. `-e ignore` can handle missing output but not missing input because none of the substeps can start.

**Comment by gaow** [comment]
Right ... so the question now is how to achieve it one way or another -- including the option of changing the way the workflow is written if that helps (as explained in the slides I sent you earlier this evening).

**Comment by BoPeng** [comment]
One way is to proceed, in the `-e ignore` mode, missing files are not checked initially. Instead it will be checked after `group_by`, and we proceed with substeps with valid `_input`.

**Comment by BoPeng** [comment]
#1329 allows `test` to run because now

```
input: dynamic(test_input_files) , group_by = 1
```
now automatically removes files that do not exist.

**Comment by BoPeng** [comment]
> One way is to proceed, in the -e ignore mode, missing files are not checked initially. Instead it will be checked after group_by, and we proceed with substeps with valid _input.

This is implemented so the workflow would execute the four groups even without the `dynamic` decorator.

**Comment by gaow** [comment]
Thank you @BoPeng now the behavior is as expected. However, can we simplify and rearrange the warning message a bit? That is, the text:

```
Failed to execute Rscript .sos/simulate_4_93141e5e.R
exitcode=1, workdir=/home/gaow/tmp/02-Dec-2019, stderr=test/simulate/simulate_5.stderr
---------------------------------------------------------------------------.
```

is best saved and shown at the very end before the entire workflow quits, so they dont hide somewhere in the middle of very long screen output that users cannot locate. 

Also shall we consolidate this line to the previous warning? eg, 

```
WARNING: Ignoring error from simulate (id=b7af26066c5f1731, index=4, output=test/simulate/simulate_5.rds):
```

**Comment by BoPeng** [comment]
@gaow Our interpretation of `keep-going` and `ignore-errors` is different from GNU Make (https://stackoverflow.com/questions/53039550/makefile-ignore-errors-vs-keep-going). I think,

1. what we are doing now with `ignore-error` is the `keep-going` of GNU make.
2. `ignore-error` should be something different, name ignore non-workflow-level errors such as when an `action` fails.

Therefore we should perhaps merge the behavior of `ignore-error` to `keep-going` and limit the use of `ignore-error` in a smaller scale. However, GNU Make allows the combined use of `-i -k` and we cannot, so things get tricky here.

I guess the problem is that I do not understand why `keep-going` does not proceed like `ignore-error` in the example of https://vatlab.github.io/sos-docs/doc/user_guide/error_handling.html  .

**Comment by BoPeng** [comment]
I would propose the removal of `-e keep-going` since it does not provide any additional feature than `-e ignore`.

**Comment by gaow** [comment]
Indeed `ignore-errors` simply means ignore non-zero return code from a system call, it seems. But if the output still dont exist, it will fail unless we also let it `keep-going`. 

It has happened to me that for some crazy C++ programs someone wrote they decide to `return 1` instead of `return 0` for the main function ... I had to write a wrapper around it for SoS to work. With `ignore-errors` it might help. (or we have a better name for it?)

> Therefore we should perhaps merge the behavior of ignore-error to keep-going and limit the use of ignore-error in a smaller scale.

I think you are saying: 

- `-i -k` in GNU Make = `-e keep-going` in SoS
- `-i` in GNU Make = `-e ignore-errors` in SoS

right?

**Comment by BoPeng** [comment]
> -i -k in GNU Make = -e keep-going in SoS
> -i in GNU Make = -e ignore-errors in SoS

Currently:

1. With `-e keep-going` we stop a brach at the step where an error happens, and we execute "safe" branches, and stops after all branches are executed or failed. Everything should be safe because all errors are caught.

2. With `-e ignore`, we ignore the fact that a step has failed and venture to the next step. which is different from the `-i` of GNU Make, and is like `-i -k`. The problem with `-e ignore` is that it can be unsafe to execute after a failed step since the step `20` will have different results with/without error in step `10`.


```
[20]
input: dynamic('*.txt')
```

So in summary
1. We execute more and more in the order of `abort`, `default`, `keep-going`, `ignore`.
2. `ignore` is not `-i` in GNU Make. more like `-i -k`.
3. We could implement GNU Make like `-i` to only ignore errors from actions and scripts, not from workflow execution, but we will need another name for `-i -k`.

**Comment by gaow** [comment]
I agree with your summary. For 3) we can call it `ignore-exception` ? for current `ignore` how about `skip`? That is, `skip` errors.

**Comment by BoPeng** [comment]
I do not want so many options but sometimes fine-controls are needed.

1. For actions we do have `allow_error`, although

```
sh: allow_error=True
  rm somefile
```
is a lot more trouble some than
```
    -rm something
```
in GNU make.

2. `keep-going` has a right balance of safety and gentle stopping, because all errors are caught. However, I feel that it is not a useful feature because the user would need to rerun the workflow anyway. I can even suggest that **we make `keep-going` the default so that SoS would by default execute all executable steps safely. They can do `-e abort` is they want a harsher stop.

3. Then we need to figure out what we are "ignoring" right now, and if we should separate into finer options.

**Comment by gaow** [comment]
> we make keep-going the default 

Great idea! We then also dont confuse it with the option in GNU Make

> For actions we do have `allow_error`

Sorry I keep forgetting that. This is good enough I think. 

To recap we'll make `keep-going` the default, then have `abort` and `ignore`? The question is whether it is worth keeping both `keep-going` and current `default`.

**Comment by BoPeng** [comment]
> The question is whether it is worth keeping both keep-going and current default.

The current `default` should be removed.

**Comment by gaow** [comment]
Agreed. Then we dont have either `keep-going` or `ignore-errors` to confuse with Make. Although maybe we can still consider using the word `skip` over `ignore`

**Comment by BoPeng** [comment]
I am glad that there is one mode to remember, especially for less confusion for differences between `default` and `keep-going` and between `keep-going` and `ignore`.

Pushed and https://vatlab.github.io/sos-docs/doc/user_guide/error_handling.html is updated. The document will be updated with more example for how `-e ignore` ignores different types of errors.

**Comment by gaow** [comment]
Nice! Before you update examples, how about the behavior of error messages, with user-friendliness (concise and easy to track) in large benchmarks in mind (https://github.com/vatlab/sos/issues/1266#issuecomment-561226798)?

**Comment by BoPeng** [comment]
Currently I see four warnings

```
WARNING: Ignoring error from substep simulate (id=b7af26066c5f1731, index=4):
Failed to execute Rscript .sos/simulate_4_8cc2f7f1.R
exitcode=1, workdir=/Users/bpeng1/sos/sos/issue, stderr=test/simulate/simulate_5.stderr
---------------------------------------------------------------------------.

WARNING: Output target test/simulate/simulate_5.rds does not exist after the completion of step a_simulate

WARNING: Output target test/test/simulate_5_test_1.rds does not exist after the completion of step a_test

WARNING: Output target test/test/simulate_5_test_1.rds does not exist after the completion of step default
```

The last two complained about the same missing file, but for different steps. The first warning message is the real message, but it is pretty difficult to move to the end because sos will need to accumulate all four messages and display them at the end, and then it makes it difficult to figure out when the errors happened if we move them out of the location they were triggered.. I would suggest a separate enhancement ticket for it.

**Comment by gaow** [comment]
> it is pretty difficult to move to the end because sos will need to accumulate all four messages and display them at the end,

I can imaging that ... 

but if we could remove the warming about missing files and put the file name into the first warning, then at least when people track for warning messages they will scroll up and get right to the real warning. Now with all 3 messages people only have 1/3 chance looking at the right one ... 

```
WARNING: Ignoring error from simulate (id=b7af26066c5f1731, index=4, output=test/simulate/simulate_5.rds):
```

(see how `output=test/simulate/simulate_5.rds` is added to the message above)

**Comment by BoPeng** [comment]
>  remove the warming about missing files '

No. We do not know what file(s) the script will generate, and if the failure will cause the missing of them.

> and put the file name into the first warning, 

These look like legitimate warning messages because the steps and workflows got an error, although in this case they are nested workflows with the same missing output file.

**Comment by BoPeng** [comment]
The confusing part could be

```
INFO: a_test output:   test/test/simulate_1_test_1.rds test/test/simulate_2_test_1.rds... (5 items)
WARNING: Output target test/test/simulate_5_test_1.rds does not exist after the completion of step a_test
```
You see the `INFO` says 5 files (in green) and then one missing. In theory they can be combined as 
```
..... (5 items, 1 missing).
```

**Comment by BoPeng** [comment]
I do not particularly like https://github.com/vatlab/sos/commit/d8fee925b5acc980ed31f9fded1ac00865cbbfbc but we can keep it if you prefer having one message instead of two.

**Comment by gaow** [comment]
I think it reads good -- I made a couple of commits to further improve it, based on what I see running `sos run test_run.sos` from this ticket. 

So we are all set now? This is very nice.

**Comment by gaow** [comment]
@BoPeng additional concern: when for `task` executions on cluster a task died silently due to wall time or memory limit on SoS level it receives a message along the lines of "task is inactive for ?? secs". In that scenario will what we've done here still help? That is with `-e ignore ` i hope to be able to skip the substeps that takes long time or lots of memory to run. That was what motivated me to bump this ticket. I can certainly test and see if we expect it to work as of now.

**Comment by BoPeng** [comment]
That is another ticket that can only be debugged on your cluster (mine has a 1 hr limit on walltime). The failed tasks will be ignored with `-e ignore` so your workflow should complete though.

**Comment by BoPeng** [comment]
@gaow I remember that you posted an example with 2 minutes jobs etc, but I cannot find it now. Could you send me a link?

**Comment by gaow** [comment]
@BoPeng sure this is another opened ticket:

https://github.com/vatlab/sos/issues/1323#issuecomment-557846407

**Comment by gaow** [comment]
I dont have a reproducible example now, but although `-e ignore` option skips failed runs the workflow hangs without ultimately completing. I'm on the cluster so maybe there is still communication issues with job scheduler when errors are ignored? I'll try to come up with MWE later but just in case there are something obvious to check now.

**Comment by BoPeng** [comment]
I saw it in one case and fixed it, but I could not get a simple enough example to be used as a testcase. Basically ignoring error without proper signaling will cause another step to wait for something. I found the problem by comparing `sos_debug.log` generated without error and with error and `-e ignore`.

**Comment by gaow** [comment]
Hmm I'm afraid there is a bigger problem here. My benchmark pipeline was completely broken after the update -- my script simply now takes in the wrong dynamic input files. I'm still investigating (by rerunning without -e ignore) if the new feature cause the issue, but what's the behavior for this:

```
input: dynamic(mnm_high_het_input_files) , group_by = 3,
```

suppose I've got `ABC`, and `DEF`. If `B` fails and was ignored, is it going to regroup as `ACD` and `EF`?

**Comment by BoPeng** [comment]
> If B fails and was ignored, is it going to regroup as ACD and EF?

Yes. This is #1329.

**Comment by BoPeng** [comment]
> If B fails and was ignored, is it going to regroup as ACD and EF?

Note that without `dynamic`, the entire `ABC` would be ignored due to missing `B`.

**Comment by gaow** [comment]
I see. this is the problem ... because it breaks the logic in [this workflow](https://github.com/gaow/random-nbs/blob/master/slides/20191201_DSC_SoS/sos/test_run.sos). I'd appreciate suggestions here (including revisiting the way `test_run.sos` is generated) ... For now I'm going to rerun without `-e ignore` (and with sos_debug.log) to verify that other recent changes don't break my workflow.

**Comment by BoPeng** [comment]
But why do you have to use `dynamic`? It is designed for `dynamic('*.txt')`, not `dynamic('output.txt')`.

**Comment by gaow** [comment]
Hmm you are saying we dont need `dynamic()` in `test_run.sos`? I was pretty sure at some point I tried without `dynamic` but it complains that input do not exist. Maybe since I also use `depends: sos_step()` then `input` don't get checked before they are actually generated by the previous step and that makes it unnecessary to use `dynamic`?

**Comment by BoPeng** [comment]
Can you try without `dynamic`? If you are sure the input should exist when the step is executed, then it should be arranged to be executed after other steps and `dynamic` is not needed. Here you use `dynamic` as a convenient way to stop SoS from running it as "head" of a branch, and we should discuss the best way to do it as a separate topic.

**Comment by gaow** [comment]
I'm trying without `dynamic` and also without `-e ignore` things seem to work. I've pushed revised version here:

https://github.com/gaow/random-nbs/blob/master/slides/20191201_DSC_SoS/sos/test_run.sos

> If you are sure the input should exist when the step is executed,

Well with `depends: sos_step` all inputs should have been generated before a step actually executes. Previously I used `dynamc` probably because I did not have `depend` on steps back then, or for whatever other reasons it compains about missing input that I dont recall. 

> we should discuss the best way to do it as a separate topic.

Will do.

**Comment by BoPeng** [comment]
Some examples crafted at https://github.com/vatlab/sos/issues/1329#issuecomment-564097025

**Comment by gaow** [comment]
I tried with `-e ignore` again. My workflow completed properly with a few corner cases identified that led to failure. After fixing those corner cases and running again SoS correctly skipped completed runs, with tasks regrouped focusing on those previously failed corner cases. It then completed. I cannot reproduce the hang behavior that was caused by messed up grouping due to the use of `dynamic` (this new run I stopped using `dynamic`)

**Comment by BoPeng** [comment]
OK. Thanks for your update. I think this would alleviate the performance problem you mentioned in #1321 . Actually, after some more thought,
```
task:
input:
```
does not achieve much since currently master tasks only saves one copy of the script anyway. The only difference is that with 
```
task:
input:
```
we are resolving substeps when the tasks are executed, and with
```
input:
task: trunk_size=0
```
we are resolving substeps for each subtask. The latter has the advantage of skipping completed substeps.

I will close this ticket for now.
